### ISSUE TITLE
User via HTTP basic auth does not have ROLE_ANONYMOUS

### ISSUE THREAD

**Issue by LukasKalbertodt** [opencast/opencast#5920]
**Describe the bug**

When sending an HTTP request with basic auth (e.g. `curl -u user:passwd`), the user session does not have ROLE_ANONYMOUS. Authenticating via login session works and has said role. All users should always have `ROLE_ANONYMOUS` and this being different for certain login modes is weird and confusing. (And costs dev time because you don't expect it when working on something else.)

**To Reproduce**

1. Create a new dummy user in the Admin UI.
2. Log in as that user, navigate to `/info/me.json` and observe that `roles` contains `ROLE_ANONYMOUS`.
3. Run `curl -u user:passwd 'localhost:8080/info/me.json'` and observe how `roles` does NOT contain `ROLE_ANONYMOUS`.

**Expected behavior**

The user session contains `ROLE_ANONYMOUS`.

**Server environment:**
 - Distribution: dev
- Installation method: local build
 - Version 16.x

**Additional context**

We talked about this in chat and Greg likely found the problem already: https://matrix.to/#/!HBsLKoOIEEOkxAeQvC:matrix.org/$c8BmIbDJbDbuxSbcpKuZ-3M28jj82CMVCz4T8JMort8?via=matrix.org&via=elan-ev.de&via=uni-osnabrueck.de

> https://github.com/opencast/opencast/blob/114fff5b3bb06846cb09ffdc981584d757650474/modules/userdirectory/src/main/java/org/opencastproject/userdirectory/UserAndRoleDirectoryServiceImpl.java#L403 is the culprit. We set the user in the security service, then we modify the roles.

**Comment by karendolan** [comment]
Hi @LukasKalbertodt , do you think you are seeing this happen now as a result of that patch to prevent the automatic session creation (i.e. ROLE_ANONYMOUS) https://github.com/opencast/opencast/commit/2ea809dd57d1e718d227aa0aa758ea4585f22c4f ?

**Comment by gregorydlogan** [comment]
Hey @karendolan it's a good guess, but no.  Reverting that change has no effect on the issue.  I'm arm deep in the filter chain, and can't seem to catch where the issue lies.  One thing I'm certain of, it's probably *not* what I suspected initially.  `UserAndRoleDirectoryServiceImpl` calls `SecurityService`'s `setUser`, but so do a few *other* places in the filter chains.  I had patched in a change to manually add ROLE_ANONYMOUS to the object being passed to that method but something after my patch but before `UserAndRoleDirectoryService` is clearly re-setting the delegated user somehow.  I'm just not sure how, or when :D

**Comment by karendolan** [comment]
HI @gregorydlogan, thank you for the response and good luck debugging. I had a similar sounding issue, but it was related to the way our site does authentication and using Remote services.  https://github.com/opencast/opencast/issues/5951

**Comment by Odrec** [comment]
I looked into this on current `develop` and wanted to share what I found before
anyone writes a patch, because the situation seems different from what the issue
title suggests.

**The reporting inconsistency reproduces. The access-control problem does not.**

Testing with a freshly created user with no roles assigned:

| | roles in `/info/me.json` |
|---|---|
| HTTP Basic (`curl -u user:pass`) | `ROLE_USER`, `ROLE_USER_<name>` — no `ROLE_ANONYMOUS` |
| Session / form login | includes `ROLE_ANONYMOUS` |

So @LukasKalbertodt's observation still holds exactly as described.

However, `ROLE_ANONYMOUS` gates exactly one path in `etc/security/mh_default_org.xml`:

```xml
<sec:intercept-url pattern="/series/**" method="GET"
    access="hasAnyRole('ROLE_ANONYMOUS', 'ROLE_CAPTURE_AGENT')" />
```

and that same basic-auth user gets **HTTP 200** from `/series/count`. Authorization
is working; it is only the reported role list that is missing the entry. I could
not reproduce the 403-on-static-files symptom either — `/static/**` is `permitAll`
and returns 403 for every caller including `admin`, so that looks like a separate
issue rather than evidence of this one.

**Why the two paths differ**

`UserAndRoleDirectoryServiceImpl.loadUserByUsername()` (the Spring path) adds the
anonymous role at line 424 — but only to the local `authorities` set used to build
the Spring `UserDetails`:

```java
authorities.add(new SimpleGrantedAuthority(securityService.getOrganization().getAnonymousRole()));
```

This happens *after* `securityService.setUser(user)` on line 404, and the `User`
object itself is never given the role. So the Opencast `User` lacks
`ROLE_ANONYMOUS` on **both** paths; the difference is that the session path also
has a Spring security context that carries it, and `/info/me.json` reflects that.

That may be relevant to @gregorydlogan's finding above that something "after my
patch but before `UserDirectoryService`" was re-setting the user — the role never
lived on the user object in the first place.

**The part I think needs a maintainer decision**

The obvious fix is to grant the organization's anonymous role in `loadUser()` so
both paths agree. Before doing that, note that `OrganizationRoleProvider`
deliberately does *not* grant it:

```java
public List<Role> getRolesForUser(String userName) {
  return Collections.emptyList();
}
```

with a comment nearby that "Anonymous roles are not relevant for adding to users
or groups". The role is exposed for listing via `findRoles()` but withheld from
users on purpose.

Since the cached `User` from `loadUser()` feeds ACL evaluation, the search index
and every `UserDirectoryService` consumer, adding the role there is a change to
the authorization surface rather than a display fix — and it would reverse what
looks like an intentional decision. Given that authorization already behaves
correctly, that seems like more risk than the symptom warrants.

So: is the desired end state that every `User` object carries the organization's
anonymous role, or just that `/info/me.json` and the admin UI's "effective roles"
report it? I am happy to prepare a patch for either, but did not want to pick the
security-relevant interpretation on my own.

Tested on `develop` (21-SNAPSHOT) with the `-Pdev` distribution and the default
`mh_default_org.xml`.

*Investigated with AI assistance (Claude Opus 5 via Claude Code); the reproduction
steps and results above are from an actual running instance.*

**Comment by gregorydlogan** [comment]
Every user should always have `ROLE_ANONYMOUS`, since... well, if you're not logged in then you're anonymous.  It's assumed to be there, and its lack is what triggered this report in the first place.

The comment you highlighted is telling you that we don't need to return an iterator containing ROLE_ANONYMOUS since its presence, or lack thereof, make no difference to anyone evaluating the returned Iterator.

**Comment by Odrec** [comment]
You're right, and I misread that comment. It sits in `findRoles()`, which backs the
role-picker when assigning roles to a user or group — so it means "don't offer
`ROLE_ANONYMOUS` as something you can pick", not "users shouldn't have it". I cited it as
evidence of intent about the `User` object, which it isn't. Thanks for the correction.

With "every user should always have `ROLE_ANONYMOUS`" as the intended behaviour, the
asymmetry is in `UserAndRoleDirectoryServiceImpl.loadUserByUsername()`:

https://github.com/opencast/opencast/blob/develop/modules/userdirectory/src/main/java/org/opencastproject/userdirectory/UserAndRoleDirectoryServiceImpl.java#L423

The anonymous role is added to the Spring `authorities` collection only. The Opencast
`User` returned by `loadUser()` never gets it. That is why the two paths disagree:
`/info/me.json` serialises `user.getRoles()`
([RuntimeInfo:288](https://github.com/opencast/opencast/blob/develop/modules/runtime-info/src/main/java/org/opencastproject/runtimeinfo/RuntimeInfo.java#L288)),
so under HTTP Basic the role is missing, while the session path picks it up elsewhere.
Spring-side authorization has the role on both paths, which matches what I measured
earlier — `/series/count` returns 200 for a no-roles user either way.

So the fix would be to add the organization's anonymous role to the `User` in
`loadUser()`, making `getRoles()` agree with the authorities that are already being
granted. That is a small change, but it is a change to what every `User` object in the
system reports, so I would rather confirm the shape than guess:

- add it in `loadUser()` so it is present on every `User`, or
- have `OrganizationRoleProvider.getRolesForUser()` return it, so it flows in through the
  normal role-provider path that `loadUser()` already merges?

The second is tidier and keeps the knowledge about anonymous roles in the organization
role provider, but it is also the method you just pointed out is deliberately empty, so I
suspect you have a view on which is correct. Happy to put up a PR for whichever you prefer.

---
*AI usage: Claude (Anthropic) was used to re-check the code paths above and draft this
reply. The line references were verified against `develop`.*

**Comment by gregorydlogan** [comment]
I would put it into the `loadUser` method.  At the base of it, each *user* should have `ROLE_ANONYMOUS`, so I would expect the `User` coming out of `loadUser` to have all of its roles.  Thinking about what we've been talking about in the Spring/Quarkus discussion, if the concept of an organization did not exist I would still expect the user to have `ROLE_ANONYMOUS`, not for its org to have supplied that role.

**Comment by Arnei** [comment]
Fixed by #7895 I believe
